test/e2e/CollateralSwap.t.sol

```diff
@@ -0,0 +1,144 @@
+
+import {AaveHelper} from "src/AaveHelper.sol";
+import {FlashLoanRouter, Loan} from "src/FlashLoanRouter.sol";
+
+import {ICowSettlement} from "src/interface/IBorrower.sol";
+
+import {Constants} from "./lib/Constants.sol";
+import {CowProtocol} from "./lib/CowProtocol.sol";
+import {CowProtocolInteraction} from "./lib/CowProtocolInteraction.sol";
+import {ForkedRpc} from "./lib/ForkedRpc.sol";
+
+contract E2eCollateralSwap is Test {
+    using ForkedRpc for Vm;
+
+    IAavePool internal constant POOL = IAavePool(0x87870Bca3F3fD6335C3F4ce8392D69350B4fA4E2);
+
+    // This is the block immediately before a mainnet fee withdrawal:
+    // <https://etherscan.io/tx/0x2ac75cbf67d74ae3ad736314acb9dba170922849d411cc7ccbe81e4e0cff157e>
+    // It guarantees that there are some WETH available in the buffers to pay
+    // for the flash loan.
+    uint256 private constant MAINNET_FORK_BLOCK = 22785269;
+    address private solver = makeAddr("E2eCollateralSwap: solver");
+    address private user = makeAddr("E2eCollateralSwap: user");
+
+    AaveBorrower private borrower;
+    FlashLoanRouter private router;
+    AaveHelper private helper;
+
+    function setUp() external {
+        vm.forkEthereumMainnetAtBlock(MAINNET_FORK_BLOCK);
+
+        // Setup flashloan router and borrower
+        router = new FlashLoanRouter(Constants.SETTLEMENT_CONTRACT);
+        borrower = new AaveBorrower(router);
+        CowProtocol.addSolver(vm, solver);
+        CowProtocol.addSolver(vm, address(router));
+
+        // Allow settlement to do approvals on borrower
+        ICowSettlement.Interaction[] memory onlyApprove = new ICowSettlement.Interaction[](1);
+        onlyApprove[0] = CowProtocolInteraction.borrowerApprove(
+            borrower, Constants.WETH, address(Constants.SETTLEMENT_CONTRACT), type(uint256).max
+        );
+        vm.prank(solver);
+        CowProtocol.emptySettleWithInteractions(onlyApprove);
+
+        // Give enough weth to our user
+        deal(address(Constants.WETH), user, 100 ether);
+        helper = new AaveHelper();
+    }
+
+    function test_10WethCollatWith100UsdsSwappingCollateralForDaiWithFlashLoan() external {
+        // Supply 10 weth, borrow 100 usds
+        vm.startPrank(user);
+        Constants.WETH.approve(address(POOL), type(uint256).max);
+        IAavePool(POOL).supply(address(Constants.WETH), 10 ether, user, 0);
+        IAavePool(POOL).borrow(address(Constants.USDS), 100 ether, 2, 0, user);
+        vm.stopPrank();
+        assertEq(Constants.AWETH.balanceOf(user), 10 ether);
+        assertEq(Constants.USDS.balanceOf(user), 100 ether);
+
+        // User approvals
+        vm.startPrank(user);
+        // Approve the settlement to pull tokens for swap
+        Constants.WETH.approve(address(Constants.SETTLEMENT_CONTRACT), type(uint256).max);
+        // Approve the helper to pull the atokens
+        Constants.AWETH.approve(address(helper), type(uint256).max);
```

Comment:
not needed for this test, but I think in practice this should become a permit and be part of the hook. This way, we can also ensure that `swap` cannot be called more than once for a user (unless they resign a new permit).